syne_tune/optimizer/schedulers/searchers/gp_fifo_searcher.py

```diff
@@ -758,9 +799,52 @@ def register_pending(
     def _fix_resource_attribute(self, **kwargs):
         pass
 
-    def _postprocess_config(self, config: dict) -> dict:
+    def _postprocess_config(self, config: Dict[str, Any]) -> Dict[str, Any]:
         return config
 
+    def _create_random_generator(self) -> CandidateGenerator:
+        if self._restrict_configurations is None:
+            return RandomStatefulCandidateGenerator(
+                hp_ranges=self._hp_ranges_for_prediction(),
+                random_state=self.random_state,
+            )
+        else:
+            return RandomFromSetCandidateGenerator(
+                base_set=self._restrict_configurations,
+                random_state=self.random_state,
+            )
+
+    def _update_restrict_configurations(
+        self,
+        new_configs: List[Dict[str, Any]],
+        random_generator: RandomFromSetCandidateGenerator,
+    ):
+        # ``random_generator`` maintains all positions returned during the
+        # search. This is used to restrict the search of ``new_configs``
+        new_ms = set(
+            self.hp_ranges.config_to_match_string(config) for config in new_configs
+        )
+        num_new = len(new_configs)
+        assert num_new >= 1
+        remove_pos = []
+        for pos in random_generator.pos_returned:
```

Comment:
The idea of this logic, involving `random_generator.pos_returned`, is to avoid having to scan over all of `self._restrict_configurations` in order to find the configs to be removed.

We remove them in order to simplify random sampling from the remaining set.